hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java

```diff
@@ -354,12 +349,7 @@ public void write(GenericRecord oldRecord) {
     if (copyOldRecord) {
       // this should work as it is, since this is an existing record
       try {
-        // rewrite file names
-        // do not preserve FILENAME_METADATA_FIELD
-        if (preserveMetadata && useWriterSchemaForCompaction) {
-          oldRecord.put(HoodieRecord.FILENAME_METADATA_FIELD_POS, newFilePath.getName());
-        }
-        fileWriter.writeAvro(key, oldRecord);
```

Comment:
in this flow, do we need to do both?
- rewriteRecord(avroRecord)
- writeAvroWithMetadata

I guess, only 2nd one would suffice. L 369. 
I understand for L296, we need both, but here we don't need to rewrite record for the case when preserveMetadata is false.

Reply:
First and foremost is that we should not be updating passed in record in place* (ie we should treat it as immutable).  Rewriting the record does essentially clone it, allowing us to modify it.

Reply:
There are obviously exceptions to this rule, but it requires guarding in making sure that the in-place changes do not escape the scope they were originally intended for